grpc.go

```diff
@@ -56,40 +58,93 @@ type onlineResource struct {
 
 var emptyResp = &gpb.Empty{}
 
-func onlineCPUMem() error {
-	resourceList := []onlineResource{
-		{
-			sysfsOnlinePath: sysfsCPUOnlinePath,
-			regexpPattern:   cpuRegexpPattern,
-		},
-		{
-			sysfsOnlinePath: sysfsMemOnlinePath,
-			regexpPattern:   memRegexpPattern,
-		},
+var onlineCPUMemLock sync.Mutex
+
+const onlineCPUMemWaitTime = 100 * time.Millisecond
+
+const onlineCPUMaxTries = 10
+
+// online resources, nbResources specifies the maximum number of resources to online.
+// If nbResources is -1 then there is no limit, all resources are connected
+// returns the number of resources connected
+func onlineResources(resource onlineResource, nbResources int32) (uint32, error) {
+	files, err := ioutil.ReadDir(resource.sysfsOnlinePath)
+	if err != nil {
+		return 0, err
 	}
 
-	for _, resource := range resourceList {
-		files, err := ioutil.ReadDir(resource.sysfsOnlinePath)
+	var count uint32
+	for _, file := range files {
+		matched, err := regexp.MatchString(resource.regexpPattern, file.Name())
 		if err != nil {
-			return err
+			return count, err
 		}
 
-		for _, file := range files {
-			matched, err := regexp.MatchString(resource.regexpPattern, file.Name())
-			if err != nil {
-				return err
-			}
+		if !matched {
+			continue
+		}
 
-			if !matched {
-				continue
-			}
+		onlinePath := filepath.Join(resource.sysfsOnlinePath, file.Name(), "online")
+		status, err := ioutil.ReadFile(onlinePath)
+		if err != nil {
+			// resource cold plugged
+			continue
+		}
 
-			onlinePath := filepath.Join(resource.sysfsOnlinePath, file.Name(), "online")
+		if strings.Trim(string(status), "\n\t ") == "0" {
 			ioutil.WriteFile(onlinePath, []byte("1"), 0600)
```

Comment:
You should probably check the error here.  What you should do I'm not sure.  Perhaps simply continue.  Looking and the old code no action was taken after this line, so although the error wasn't discarded, ignoring it was harmless.  But with these changes I guess it's not.

Reply:
fixed, thanks

Reply:
I'm not sure it's fixed correctly though.  If you get one write error you'll give up entirely, i.e., the for loop in onlineCPUResources will terminate. Is this intentional?